torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -1517,6 +1526,70 @@ def test_call_method_on_rref(self):
 
         self.assertEqual(result, sum(vals))
 
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_single_threaded_rref_owner(self):
+        # This test aims to verify if the server can handle all internal RPC 
+        # messages using just one thread. 
+        caller_rank = 0
+        callee_rank = 1
+        rpc_backend_options = rpc.ProcessGroupRpcBackendOptions(
+            init_method=self.rpc_backend_options.init_method,
+            num_send_recv_threads=1
+        ) if self.rank == callee_rank else self.rpc_backend_options
+
+        rpc.init_rpc(
+            name=worker_name(self.rank),
+            backend=self.rpc_backend,
+            rank=self.rank,
+            world_size=self.world_size,
+            rpc_backend_options=rpc_backend_options,
+        )
+
+        if self.rank == caller_rank:
+            dst = worker_name(callee_rank)
+            rrefs = []
+
+            # makes sure there is no existing OwnerRRefs on dst
+            info = rpc.rpc_sync(dst, get_rref_debug_info)
+            self.assertEqual(0, int(info["num_owner_rrefs"]))
+
+            # creating RRefs on dst
+            for i in range(20):
+                rrefs.append(
+                    rpc.remote(dst, delayed_add, args=(torch.zeros(2, 2), i))
+                )
+
+            # using RRefs on dst
+            futs = []
+            for i in range(len(rrefs)):
+                futs.append(
+                    rpc.rpc_async(dst, my_rref_function, args=(rrefs[i], rrefs[i]))
```

Comment:
I'm guessing the testing strategy is the same as the diff above in this stack - at least in some cases, we hope to get the case where we run this `rpc_async`, hence requiring unpickling of the RRef on the owner node, before we process the PYTHON_REMOTE_CALL which previously would need to be processed first. I guess if we run this test enough times without flakiness and it didn't work before this PR, then the test is fine.

BTW, do we need to test this for Torchscript functions as well? It's a little bit harder because we cannot do "delay add". Hopefully https://github.com/pytorch/pytorch/issues/36272 well help us increase this sort of test coverage when we add it.

Reply:
Will test TorchScript functions in a followup PR.